accounting/accounting_test.go

```diff
@@ -0,0 +1,1328 @@
+)
+
+func TestExchange(t *testing.T) {
+	// All possible minor currency units sampled from the currency
+	// package are in use in this suite.
+	//
+	// factors of 1, 100, 1000 and 10000 are represented.
+	//
+	// As there is no point in duplicating tests for currencies with
+	// the exact same minor unit, the better approach is to instead
+	// bulk up the amount of unique values tested.
+	//
+	// The test is essentially:
+	// value / factor / rate = want
+	//
+	// Each value in the want array has been calculated manually.
+	tests := []struct {
+		name   string
+		money  currency.Currency
+		values [4]float64
+		rates  [4]float64
+		want   [4]float64
+	}{
+		{
+			name:   "Japanese yen, factor 1",
+			money:  currency.JPY,
+			values: [4]float64{100, 1000, 10000, 1999},
+			rates:  [4]float64{1, 1.5, 145.961337, 145.961337},
+			want:   [4]float64{100, 666.67, 68.51, 13.70},
+		},
+		{
+			name:   "Zimbabwean dollar, factor 100",
+			money:  currency.ZWL,
+			values: [4]float64{100, 1000, 10000, 19999.99},
+			rates:  [4]float64{1, 1.5, 469.167, 469.167},
+			want:   [4]float64{1, 6.67, 0.21, 0.43},
+		},
+		{
+			name:   "Tunisian dinar, factor 1000",
+			money:  currency.TND,
+			values: [4]float64{100, 1000, 10000, 1999.99},
+			rates:  [4]float64{1, 1.5, 3.714896, 3.714896},
+			want:   [4]float64{0.1, 0.67, 2.69, 0.54},
+		},
+		{
+			name:   "Uruguayan centésimos, factor 10000",
+			money:  currency.UYW,
+			values: [4]float64{100, 1000, 10000, 1999999.99},
+			rates:  [4]float64{1, 1.5, 42.530333, 42.530333},
+			want:   [4]float64{0.01, 0.07, 0.02, 4.70},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			for i := 0; i < len(tt.values); i++ {
+				res := accounting.Exchange(tt.money, tt.values[i], tt.rates[i])
+				if !cmp.Equal(tt.want[i], res) {
+					t.Error(cmp.Diff(tt.want[i], t))
+				}
+			}
+		})
+	}
+}
+
+func TestRatNetAmount(t *testing.T) {
+	br := func(f64 float64) *big.Rat {
+		return new(big.Rat).SetFloat64(f64)
+	}
+
+	type args struct {
+		value *big.Rat
+		rate  *big.Rat
+	}
+	var tests = []struct {
+		name    string
+		args    args
+		want    string
+		wantErr bool
+	}{
+		{
+			name: "gross 10 vat 19",
+			args: args{
+				value: br(10),
+				rate:  br(.19),
+			},
+			want:    "8.403361345",
+			wantErr: false,
+		},
+		{
+			name: "gross 19.99 vat 20",
+			args: args{
+				value: br(19.99),
+				rate:  br(.20),
+			},
+			want:    "16.658333333",
+			wantErr: false,
+		},
+		{
+			name: "gross 123 vat 7",
+			args: args{
+				value: br(123),
+				rate:  br(.07),
+			},
+			want:    "114.953271028",
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := accounting.RatNetAmount(tt.args.value, tt.args.rate)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("RatNetAmount() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !cmp.Equal(got.Text('f', 9), tt.want) {
+				t.Error(cmp.Diff(got.Text('f', 9), tt.want))
```

Comment:
The test errors may be confusing without displaying which direction they are diffing. https://github.com/golang/go/wiki/TestComments#print-diffs.

Also, some tests are diffing want, got and some are got, want. Doesnt matter, technically, but will certainly contribute to confusion when interpreting the output.

Reply:
fixed